File: landing-page/examples/ruby/fraud-page.rb

# gem install neo4j-core
require 'neo4j-core'
session = Neo4j::Session.open(:server_db, 'http://localhost:7474', basic_auth: {username: 'neo4j', password: '<password>'})

# Insert data
insert_query = """
CREATE (hank:Person {name:'Hank'}),
(abby:Person {name:'Abby'}),
(max:Person {name:'Max'}),
(sophie:Person {name: 'Sophie'}),
(jane:Person {name: 'Jane'}),
(bill:Person {name: 'Bill'}),
(ssn993632634:SSN {number: 993632634, name: 'SSN 993632634'}),
(ssn123456789:SSN {number: 123456789, name: 'SSN 123456789'}),
(ssn523252364:SSN {number: 523252364, name: 'SSN 523252364'}),
(chase:Account {bank: 'Chase', number: 1523, name: 'Chase 1523'}),
(bofa:Account {bank: 'Bank of America', number: 4634, name: 'BofA 4634'}),
(cayman:Account {bank: 'Cayman', number: 863, name: 'Cayman 863'}),
(bill)-[:HAS_SSN]->(ssn523252364),
(bill)-[:HAS_ACCOUNT]->(bofa),
(jane)-[:HAS_SSN]->(ssn123456789),
(jane)-[:HAS_ACCOUNT]->(chase),
(hank)-[:HAS_ACCOUNT]->(cayman),
(abby)-[:HAS_ACCOUNT]->(cayman),
(abby)-[:HAS_SSN]->(ssn993632634),
(sophie)-[:HAS_SSN]->(ssn993632634),
(max)-[:HAS_SSN]->(ssn993632634)
"""
session.query(insert_query)

# Transitive Closure

transitive_query = """
MATCH (n:Person)-[*]-(o)
WHERE n.name = {name}
RETURN DISTINCT o.name AS other
"""

response = session.query(transitive_query, name: 'Hank')
puts "Suspicious entities: "
response.each do |row|
  puts row[:other]
end
puts "---------------------"
# Investigation Targeting

targeting_query = """
MATCH (n:Person)-[*]-(o)
WITH n, count(DISTINCT o) AS size
WHERE size > 2
RETURN n.name AS target
"""

response = session.query(targeting_query)
puts "Investigation targets: "
response.each do |row|
  puts row[:target]
end
puts "----------------------"

# Fast Insights

connecting_paths_query = """
MATCH (ssn:SSN)<-[:HAS_SSN]-(:Person)-[:HAS_ACCOUNT]->(acct:Account)
WHERE ssn.number = 993632634
RETURN acct.bank + ' ' + str(acct.number) AS account
"""

response = session.query(connecting_paths_query, name1: 'Joe', name2: 'Billy')
puts "Accounts: "
response.each do |row|
  puts row[:account]
end

File: landing-page/examples/dotnet/fraud-page.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Neo4jClient;
using Neo4jClient.Cypher;
using Newtonsoft.Json;

void Main()
{
    var gc = new GraphClient(new Uri("http://localhost.:7474/db/data"));
    gc.Connect();

    const string createCypher = @"(hank:Person {name:'Hank'}),
(abby:Person {name:'Abby'}),
(max:Person {name:'Max'}),
(sophie:Person {name: 'Sophie'}),
(jane:Person {name: 'Jane'}),
(bill:Person {name: 'Bill'}),
(ssn993632634:SSN {number: 993632634}),
(ssn123456789:SSN {number: 123456789}),
(ssn523252364:SSN {number: 523252364}),
(chase:Account {bank: 'Chase', number: 1523}),
(bofa:Account {bank: 'Bank of America', number: 4634}),
(cayman:Account {bank: 'Cayman', number: 863}),
(bill)-[:HAS_SSN]->(ssn523252364),
(bill)-[:HAS_ACCOUNT]->(bofa),
(jane)-[:HAS_SSN]->(ssn123456789),
(jane)-[:HAS_ACCOUNT]->(chase),
(hank)-[:HAS_ACCOUNT]->(cayman),
(abby)-[:HAS_ACCOUNT]->(cayman),
(abby)-[:HAS_SSN]->(ssn993632634),
(sophie)-[:HAS_SSN]->(ssn993632634),
(max)-[:HAS_SSN]->(ssn993632634)";
    gc.Cypher.Create(createCypher).ExecuteWithoutResults();

    //Transitive Closure
    var transitiveClosure = gc.Cypher
    	.Match("(n:Person)-[*]-(o)")
    	.Where((Entity n) => n.Name == "Hank")
    	.Return(o => o.As<Entity>())
    	.Results;
    
    Console.WriteLine("Transitive Closure");
    foreach (var tc in transitiveClosure)
    	Console.WriteLine($"\t{tc}");

    //Investigation Targeting
    var investigationTargeting = gc.Cypher
    	.Match("(n:Person)-[*]-(o)")
    	.With("n, count(distinct o) AS size")
    	.Where("size > 2")
    	.Return(n => n.As<Entity>())
    	.Results;

    Console.WriteLine("Investigation Targeting");
    foreach (var it in investigationTargeting)
    	Console.WriteLine($"\t{it}");
    
    //Fast Insights
    var fastInsights = gc.Cypher
    	.Match("(ssn:SSN)<-[:HAS_SSN]-(:Person)-[:HAS_ACCOUNT]->(acct:Account)")
    	.Where((Entity ssn) => ssn.Number == 993632634)
    	.Return(acct => acct.As<Entity>())
    	.Results;
    
    Console.WriteLine("Fast Insights");
    foreach (var entity in fastInsights)
    	Console.WriteLine($"\tBank: {entity.Bank}, Number: {entity.Number}");
}

public class Entity
{
    [JsonProperty("name")]
    public string Name { get; set; }

    [JsonProperty("bank")]
    public string Bank { get; set;}

    [JsonProperty("number")]
    public int Number { get; set;}

    public override string ToString()
    {
    	return (string.IsNullOrWhiteSpace(Name)) ? ((string.IsNullOrWhiteSpace(Bank))) ?  Number.ToString() : Bank : Name;
    }
}
